# Module 1.

## Pre-processing the dataset:

### In this task you will be working on handling null values, deletion or transformation of irrelevant values, data type transformation, removing duplicates and data validations. Once you have completed this task, you will get a refined and a cleaner data set for further analysis.

#### Steps to perform Data Pre-processing: 

#### Step 1: Removing duplicate rows ( there could be duplicate rows excluding Row_ID column ).

#### Step 2: Removing rows for which few values are missing.

#### Step 3: Remove irrelevant values from each column if any. Validation of all values for a column( order date and ship date value must be in correct date format ). For each entry in dataset ship date >= order date

#### Step 4: Export the cleaned dataset as a .csv file: prefer UTF-8 encoding.

In [1]:
#importing required libraries
import numpy as np
import pandas as pd

In [2]:
#load data
store_data = pd.read_csv(r'C:\Users\Danish\Downloads\superstore_final_dataset.csv')

In [3]:
#displayong the dataframe
store_data.head()

Row_ID        Order_ID  Order_Date   Ship_Date       Ship_Mode Customer_ID  \
0       1  CA-2017-152156  08/11/2017  11/11/2017    Second Class    CG-12520   
1       2  CA-2017-152156  08/11/2017  11/11/2017    Second Class    CG-12520   
2       3  CA-2017-138688  12/06/2017  16/06/2017    Second Class    DV-13045   
3       4  US-2016-108966  11/10/2016  18/10/2016  Standard Class    SO-20335   
4       5  US-2016-108966  11/10/2016  18/10/2016  Standard Class    SO-20335   

     Customer_Name    Segment        Country             City       State  \
0      Claire Gute   Consumer  United States        Henderson    Kentucky   
1      Claire Gute   Consumer  United States        Henderson    Kentucky   
2  Darrin Van Huff  Corporate  United States      Los Angeles  California   
3    Sean O Donnel   Consumer  United States  Fort Lauderdale     Florida   
4    Sean O Donnel   Consumer  United States  Fort Lauderdale     Florida   

   Postal_Code Region       Product_ID         Category Sub_Category  \
0      42420.0  South  FUR-BO-10001798        Furniture    Bookcases   
1      42420.0  South  FUR-CH-10000454        Furniture       Chairs   
2      90036.0   West  OFF-LA-10000240  Office Supplies       Labels   
3      33311.0  South  FUR-TA-10000577        Furniture       Tables   
4      33311.0  South  OFF-ST-10000760  Office Supplies      Storage   

                                        Product_Name     Sales  
0                  Bush Somerset Collection Bookcase  261.9600  
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...  731.9400  
2  Self-Adhesive Address Labels for Typewriters b...   14.6200  
3      Bretford CR4500 Series Slim Rectangular Table  957.5775  
4                      Eldon Fold N Roll Cart System   22.3680

In [4]:
#info
store_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9800 entries, 0 to 9799
Data columns (total 18 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Row_ID         9800 non-null   int64  
 1   Order_ID       9800 non-null   object 
 2   Order_Date     9800 non-null   object 
 3   Ship_Date      9800 non-null   object 
 4   Ship_Mode      9800 non-null   object 
 5   Customer_ID    9800 non-null   object 
 6   Customer_Name  9800 non-null   object 
 7   Segment        9800 non-null   object 
 8   Country        9800 non-null   object 
 9   City           9800 non-null   object 
 10  State          9800 non-null   object 
 11  Postal_Code    9789 non-null   float64
 12  Region         9800 non-null   object 
 13  Product_ID     9800 non-null   object 
 14  Category       9800 non-null   object 
 15  Sub_Category   9800 non-null   object 
 16  Product_Name   9800 non-null   object 
 17  Sales          9800 non-null   float64
dtypes: float

In [5]:
store_data.isnull().sum()

Row_ID            0
Order_ID          0
Order_Date        0
Ship_Date         0
Ship_Mode         0
Customer_ID       0
Customer_Name     0
Segment           0
Country           0
City              0
State             0
Postal_Code      11
Region            0
Product_ID        0
Category          0
Sub_Category      0
Product_Name      0
Sales             0
dtype: int64

#### Only postal_code has 11 null values out of 9800 rows.  

In [6]:
store_data.describe().transpose()

count          mean           std       min        25%  \
Row_ID       9800.0   4900.500000   2829.160653     1.000   2450.750   
Postal_Code  9789.0  55273.322403  32041.223413  1040.000  23223.000   
Sales        9800.0    230.769059    626.651875     0.444     17.248   

                  50%        75%       max  
Row_ID        4900.50   7350.250   9800.00  
Postal_Code  58103.00  90008.000  99301.00  
Sales           54.49    210.605  22638.48

### Step 1: Removing duplicate rows ( there could be duplicate rows excluding Row_ID column ). 

In [7]:
#duplicate data removal
ss_data = store_data.drop_duplicates()

In [8]:
#Reading the new dataframe
ss_data.head()

Row_ID        Order_ID  Order_Date   Ship_Date       Ship_Mode Customer_ID  \
0       1  CA-2017-152156  08/11/2017  11/11/2017    Second Class    CG-12520   
1       2  CA-2017-152156  08/11/2017  11/11/2017    Second Class    CG-12520   
2       3  CA-2017-138688  12/06/2017  16/06/2017    Second Class    DV-13045   
3       4  US-2016-108966  11/10/2016  18/10/2016  Standard Class    SO-20335   
4       5  US-2016-108966  11/10/2016  18/10/2016  Standard Class    SO-20335   

     Customer_Name    Segment        Country             City       State  \
0      Claire Gute   Consumer  United States        Henderson    Kentucky   
1      Claire Gute   Consumer  United States        Henderson    Kentucky   
2  Darrin Van Huff  Corporate  United States      Los Angeles  California   
3    Sean O Donnel   Consumer  United States  Fort Lauderdale     Florida   
4    Sean O Donnel   Consumer  United States  Fort Lauderdale     Florida   

   Postal_Code Region       Product_ID         Category Sub_Category  \
0      42420.0  South  FUR-BO-10001798        Furniture    Bookcases   
1      42420.0  South  FUR-CH-10000454        Furniture       Chairs   
2      90036.0   West  OFF-LA-10000240  Office Supplies       Labels   
3      33311.0  South  FUR-TA-10000577        Furniture       Tables   
4      33311.0  South  OFF-ST-10000760  Office Supplies      Storage   

                                        Product_Name     Sales  
0                  Bush Somerset Collection Bookcase  261.9600  
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...  731.9400  
2  Self-Adhesive Address Labels for Typewriters b...   14.6200  
3      Bretford CR4500 Series Slim Rectangular Table  957.5775  
4                      Eldon Fold N Roll Cart System   22.3680

In [9]:
ss_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9800 entries, 0 to 9799
Data columns (total 18 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Row_ID         9800 non-null   int64  
 1   Order_ID       9800 non-null   object 
 2   Order_Date     9800 non-null   object 
 3   Ship_Date      9800 non-null   object 
 4   Ship_Mode      9800 non-null   object 
 5   Customer_ID    9800 non-null   object 
 6   Customer_Name  9800 non-null   object 
 7   Segment        9800 non-null   object 
 8   Country        9800 non-null   object 
 9   City           9800 non-null   object 
 10  State          9800 non-null   object 
 11  Postal_Code    9789 non-null   float64
 12  Region         9800 non-null   object 
 13  Product_ID     9800 non-null   object 
 14  Category       9800 non-null   object 
 15  Sub_Category   9800 non-null   object 
 16  Product_Name   9800 non-null   object 
 17  Sales          9800 non-null   float64
dtypes: float

#### There are no duplicate values

### Step 2: Removing rows for which few values are missing.

#### Only postal code has 11 missing values. Lets remove it as required.

In [10]:
#dropping columns
ss_data.drop('Postal_Code', axis=1, inplace=True)

In [11]:
ss_data.columns

Index(['Row_ID', 'Order_ID', 'Order_Date', 'Ship_Date', 'Ship_Mode',
       'Customer_ID', 'Customer_Name', 'Segment', 'Country', 'City', 'State',
       'Region', 'Product_ID', 'Category', 'Sub_Category', 'Product_Name',
       'Sales'],
      dtype='object')

### Step 3: Remove irrelevant values from each column if any. Validation of all values for a column( order date and ship date value must be in correct date format ). For each entry in dataset ship date >= order date 

In [12]:
# Convert dates to datetime format
ss_data['Ship_Date'] = pd.to_datetime(ss_data['Ship_Date'])
ss_data['Order_Date'] = pd.to_datetime(ss_data['Order_Date'])

C:\Users\Danish\anaconda3\lib\site-packages\pandas\core\tools\datetimes.py:1047: UserWarning: Parsing '16/06/2017' in DD/MM/YYYY format. Provide format or specify infer_datetime_format=True for consistent parsing.
  cache_array = _maybe_cache(arg, format, cache, convert_listlike)
C:\Users\Danish\anaconda3\lib\site-packages\pandas\core\tools\datetimes.py:1047: UserWarning: Parsing '18/10/2016' in DD/MM/YYYY format. Provide format or specify infer_datetime_format=True for consistent parsing.
  cache_array = _maybe_cache(arg, format, cache, convert_listlike)
C:\Users\Danish\anaconda3\lib\site-packages\pandas\core\tools\datetimes.py:1047: UserWarning: Parsing '14/06/2015' in DD/MM/YYYY format. Provide format or specify infer_datetime_format=True for consistent parsing.
  cache_array = _maybe_cache(arg, format, cache, convert_listlike)
C:\Users\Danish\anaconda3\lib\site-packages\pandas\core\tools\datetimes.py:1047: UserWarning: Parsing '20/04/2018' in DD/MM/YYYY format. Provide format or sp

In [13]:
# Create a new column 'ship_after_order' that is True if the ship date is >= order date, otherwise False
ss_data['ship_after_order'] = ss_data['Ship_Date'] >= ss_data['Order_Date']

In [14]:
#checking unique values
ss_data['ship_after_order'].unique()

array([ True, False])

#### The above result shows that there are rows where ship date is < order date which is absurd. 

In [15]:
# Get the count of unique values in a column
unique_count = ss_data['ship_after_order'].nunique()

In [16]:
# Get the count of how many times a value appears in a column
value_count = ss_data['ship_after_order'].value_counts()

In [17]:
# Print the value counts
print(value_count)

True     8116
False    1684
Name: ship_after_order, dtype: int64


#### It can be clearly observed that there are 1684 instances where the data for ship date and order date is absurd.

#### These rows need to be removed because it contains absurd/irrelevant/wrong values.

In [20]:
# Drop rows where a specific column value equals a certain value
df = ss_data.drop(ss_data[ss_data['ship_after_order'] == False].index)

In [21]:
# Print the updated DataFrame
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8116 entries, 0 to 9795
Data columns (total 18 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   Row_ID            8116 non-null   int64         
 1   Order_ID          8116 non-null   object        
 2   Order_Date        8116 non-null   datetime64[ns]
 3   Ship_Date         8116 non-null   datetime64[ns]
 4   Ship_Mode         8116 non-null   object        
 5   Customer_ID       8116 non-null   object        
 6   Customer_Name     8116 non-null   object        
 7   Segment           8116 non-null   object        
 8   Country           8116 non-null   object        
 9   City              8116 non-null   object        
 10  State             8116 non-null   object        
 11  Region            8116 non-null   object        
 12  Product_ID        8116 non-null   object        
 13  Category          8116 non-null   object        
 14  Sub_Category      8116 n

In [24]:
#verifying data
df['ship_after_order'].unique()

array([ True])

### Step 4: Export the cleaned dataset as a .csv file: prefer UTF-8 encoding. 

In [25]:
# Export data to a CSV file with UTF-8 encoding
df.to_csv('cleaned_superstore_data.csv', encoding='utf-8', index=False)